feat(vim.pack): lockfile support

echasnovski opened the pull request:
This PR adds initial lockfile support to `vim.pack`. Resolves #34776.

TL;DR (more details are in commit messages):

- Add tracking of relevant data via on disk file. It is located at '$XDG_CONFIG_HOME/nvim/nvim-pack.lock' and is recommended to be put under version control (as the rest of the config). **Edit**: it was decided to go with 'nvim-pack-lock.json' as a name.

    The file is read exactly once during an active session (for performance) and is updated (both in memory and on disk) only when needed.

- Update several functions to use lockfile internally:
    - `get()` takes `version` of non-active plugin from lockfile.
    - `get()` always returns plugin's `rev` regardless of `opts.info` (as it now doesn't need Git CLI call).
    - `update()` now updates all plugins by default, even non-active (because their `version` is now known).

- Prefer using revision from lockfile when installing a plugin, instead of pulling latest according to `version`. This seems to be the most expected side effect of present lockfile.

Notes:

- **Important**. The lockfile name is 'nvim-pack.lock', which doesn't make it immediately obvious that it is a JSON file. It will be detected like one within Neovim, but not outside. Since the file needs to contain 'nvim-pack' and 'lock' somewhere in its name, I decided to go with current name. The alternative is 'nvim-pack-lock.json', but it might be too verbose. If it is deemed not, I'd rather go with it. So this requires executive decision.

- The current implementation should work reliably when `vim.pack` is used as expected, but *might* be a bit brittle when it comes to unexpected situations. For example:

    - Lockfile is manually edited or deleted entirely. It is explicitly documented to not do that. To further mitigate that, I plan to add `:checkhealth` support for lockfile also (#34777) which will contain suggestions on how to fix situations.
    - There is an error in between state change and lockfile write. This doesn't seem likely based on the code, but not 100% sure.
    - There are several Neovim instances that try to write to the same lockfile. Not sure how to mitigate this without much complications. I guess just saying "please don't do that" should be enough.

- This PR does not contain all planned usages of lockfile data. I also plan to update `vim.pack.add()` to perform extra actions (#34771 and maybe ensuring installation of all plugins from lockfile), but those will probably require more discussions and this PR is already big enough.

**Edit**: Several hours after posting I realized that I made a demo and forgot to post here :facepalm: Here it is:

https://github.com/user-attachments/assets/58815610-16e4-4be0-a562-344953b61370

echasnovski commented:
> Does it really need to include 'nvim'? What about `pack-lock.json` or even without hyphen `packlock.json`. Or if `.json` extension isn't essential `pack.lock` feels pretty clean.

The original chain of reasoning is as follows:
- The plugin manager functions are put into `vim.pack` to have a nice "`vim.pack.add` is a smarter `:packadd`". It also has to be in `vim.xxx` namespace, as all Neovim's Lua functions go there.
- It requires own names for buffer filetype and URI name prefix. The "pack" seems too expansive, the "vim.pack" implies that it comes from Vim, "nvim.pack" contains problematic dot. So "nvim-pack" it is for both for more consistency.
- The log file is named "nvim-pack.log" also in favor of consistency with filetype and buffer names. Although I agree that 'nvim-' here might be redundant since it is placed in the Nvim specific directory.
- So in favor of consistency, the lockfile is named 'nvim-pack.lock'.

I am generally not really opposed to 'pack.lock' or 'pack-lock.json, but only if the log file is also renamed to 'pack.log'. Although inconsistency with filetype and buffer name is a bit annoying.

skewb1k commented:
> I am generally not really opposed to 'pack.lock' or 'pack-lock.json, but only if the log file is also renamed to 'pack.log'.

Consistency with the log file makes sense, but the 'nvim-' prefix feels like overhead, especially considering that no other log file has it.

> Although inconsistency with filetype and buffer name is a bit annoying.

That's part of why I think it's worth considering to drop the 'nvim-' prefix to avoid ending up with something huge like `nvim-pack-lock.json`.

yochem commented:
> Add tracking of relevant data via on disk file. It is located at '$XDG_CONFIG_HOME/nvim/nvim-pack.lock' and is recommended to be put under version control (as the rest of the config).

I understand the intend of this (to make it easy for users to VC it), but it doesn't sit right to me to break the XDG spec for it. The lockfile is in no means configuration, but would fall under the data category. I would much rather see `$XDG_DATA_HOME/nvim/nvim-pack.lock` as default path, which can be configured or symlinked if configuration is out of the question. 

Also, many dotfiles approaches allow for tracking files in the data dir too. For example the popular git --bare works just fine and I'm sure gnu stow would also work with this.

yochem commented:
> The way I see it is that such lockfile contains information relevant to the config, in particular "latest working revision". The only difference is that it is managed automatically and not manually.

Still seems like data to me. The same could be said about plugins added by vim pack. They also "configure" the editor. E.g. I don't really care about locking the (few) plugins that I have and just follow the main branch. This is already encoded in my configuration by not specifying a `version` to the specs. I don't "configure" Nvim to checkout exactly commit abcdef of pluginX. 

> Although true, designing the default lockfile location around the idea of bare Git repositories is a bit far fetched.

That was not my point. My point is that currently the design is based on the assumption that it is important to put the lock file under version control and that therefore the XDG config directory is a better fit, which I think is the wrong way to decide where such a file should live. 

Anyway, that was my 2c. As long as it's configurable (or can be disabled) I'm fine with it.

To end on a more positive note, Thanks for `vim.pack` and your careful design decisions around it. I love it already and am glad to finally ditch the external package managers :)

justinmk commented:
> Still seems like data to me. The same could be said about plugins added by vim pack.

lockfiles generally are intended to be tracked in VCS, whereas plugins are not. The lockfile represents the plugin state, thus the plugins themselves are just the "uncompressed" state.

However, we can always revisit if needed. Maybe 

> alternative is 'nvim-pack-lock.json', but it might be too verbose

`nvim-pack-lock.json` seems fine, and a .json extension definitely seems like the right choice. Prepending Nvim-owned things with `nvim.` or `nvim-` is a good idea in general, wherever we share a namespace with non-Nvim things.

justinmk reviewed (approved):
LGTM, I would merge it but probably should settle the filenames first:

- `nvim-pack.log` name seems good to me.
- `nvim-pack-lock.json` seems fine, and a .json extension definitely seems like the right choice. 
    - Prepending Nvim-owned things with `nvim.` or `nvim-` is generally required wherever we share a namespace with non-Nvim things (e.g. 3P plugins). (The default `$NVIM_LOG_FILE` actually should be renamed.)

justinmk commented:
> both `vim.pack` files (log and lock) are located in Nvim related directory

? `vim.fn.stdpath('config')` is shared with all plugins.

echasnovski commented:
> > both `vim.pack` files (log and lock) are located in Nvim related directory
> 
> ? `vim.fn.stdpath('config')` is shared with all plugins. Not all files in there are Nvim-owned (i.e. core Nvim).

My bad. I misunderstood "with non-Nvim things" as "things that are not related to Neovim at all" and not "Neovim core".

Yeah, making extra clear in the name that it is owned by Nvim is a good thing.

echasnovski commented on runtime/doc/pack.txt:
```diff
@@ -221,6 +221,13 @@ Uses Git to manage plugins and requires present `git` executable of at least
+The latest state of all managed plugins is stored inside a *vim.pack-lockfile*
+located at `$XDG_CONFIG_HOME/nvim/nvim-pack.lock`. It is a JSON file that is
+used to persistently track data about plugins. Put it under version control
+for a more robust config, as initial install prefers revision from the
+lockfile instead of inferring from `version`. Should not be edited by hand or
```
I don't think this is needed:

- If user wants the initial install based on lockfile revision - treat lockfile as part of the config. Like add to VCS and have it present for the initial install.
- If user doesn't want that - don't treat lockfile like part of the config. If it is not present on initial install, then `version` is used.

echasnovski committed:
feat(pack): add initial lockfile tracking

Problem: Some use cases require or benefit from persistent on disk
  storage of plugin data (a.k.a. "lockfile"):

  1. Allow `update()` to act on not-yet-active plugins. Currently if
    `add()` is not yet called, then plugin's version is unknown
    and `update()` can't decide where to look for changes.

  2. Efficiently know plugin's dependencies without having to read
    'pkg.json' files on every load for every plugin. This is for the
    future, after there is `packspec` support (or other declaration of
    dependencies on plugin's side).

  3. Allow initial install to check out the exact latest "working" state
    for a reproducible setup. Currently it pulls the latest available
    `version.`

  4. Ensure that all declared plugins are installed, even if lazy loaded.
    So that later `add()` does not trigger auto-install (when there
    might be no Internet connection, for example) and there is no issues
    with knowing which plugins are used in the config (so even never
    loaded rare plugins are still installed and can be updated).

  5. Allow `add()` to detect if plugin's spec has changed between
    Nvim sessions and act accordingly. I.e. either set new `src` as
    origin or enforce `version.` This is not critical and can be done
    during `update()`, but it might be nice to have.

Solution: Add lockfile in JSON format that tracks (adds, updtes,
  removes) necessary data for described use cases. Here are the required
  data that enables each point:

    1. `name` -> `version` map.
    2. `name` -> `dependencies` map.
    3. `name` -> `rev` map. Probably also requires `name` -> `src` map
       to ensure that commit comes from correct origin.
    4. `name` -> `src` map. It would be good to also track the order,
       but that might be too many complications and redundant together
       with point 2.
    5. Map from `name` to all relevant spec fields. I.e. `name` -> `src`
       and `name` -> `version` for now. Storing data might be too much,
       but can be discussed, of course.

  This commit only adds lockfile tracking without implementing actual
  use cases. It is stored in user's config directory and is suggested to
  be tracked via version control.

  Example of a lockfile:

  ```json
  {
    # Extra nesting to more future proof.
    "plugins": {
      "plug-a": {
        "ref": "abcdef1"
        "src": "https://github.com/user/plug-a",
        # No `version` means it was `nil` (infer default branch later)
      },
      "plug-b": {
        "dependencies": ["plugin-a", "plug-c"],
        "src": "https://github.com/user/plug-b",
        "ref": "bcdefg2",
        # Enclose string `version` in quotes
        "version": "'dev'"
      },
      "plug-c": {
        "src": "https://github.com/user/plug-c",
        "ref": "cdefgh3",
        # Store `vim.version.Range` via its `tostring()` output
        "version": ">=0.0.0",
      }
    }
  }
  ```

echasnovski committed:
fix(pack)!: use lockfile in `get()` for data about non-active plugins

Problem: `get()` doesn't return `spec.version` about not-yet-active
  plugins (because there was no way to know that without `add()`).

Solution: Use lockfile data to set `spec.version` of non-active plugins.

echasnovski committed:
feat(pack)!: update `get()` to return revision regardless of `opts.info`

Problem: The revision data is returned behind `opts.info` flag because
  it required extra Git calls. With lockfile it is not the case.

Solution: Use lockfile to always set `rev` field in output of `get()`.

echasnovski committed:
feat(pack)!: make `update()` include not active plugins by default

Problem: Running `update()` by default doesn't include not active
  plugins, because there was no way to get relevant `version` to get
  updates from. This might be a problem in presence of lazy loaded
  plugins, i.e. ones that can be "not *yet* active" but still needed to
  be updated.

Solution: Include not active plugins by default since their `version` is
  tracked via lockfile.

echasnovski committed:
feat(pack): prefer using revision from lockfile during install

Problem: Installing plugin always pulls latest `version` changes
  (usually from the default branch or "latest version tag"). It is more
  robust to prefer initial installation to use the latest recorded
  (i.e. "working") revision.

Solution: Prefer using revision from the lockfile (if present) during
  install. The extra `update()` will pull the latest changes.

echasnovski commented:
Pushed an update:
- Rebased on latest `master`.
- Renamed lockfile to 'nvim-pack-lock.json'.
- Reworded lockfile recommendation to be more general "For a more robust config treat lockfile like its part: put under version control, etc." instead of simply suggesting to put under version control.
- Minor tweaks to example workflows.

justinmk commented on test/functional/plugin/pack_spec.lua:
```diff
@@ -413,6 +430,117 @@ describe('vim.pack', function()
+        '    "basic": {',
+        '      "rev": "' .. basic_rev .. '",',
+        '      "src": "' .. repos_src.basic .. '",',
+        -- Branch, tag, and commit should be serialized like `'value'` to be
+        -- distinguishable from version ranges
+        '      "version": "\'some-tag\'"',
```
Hmm...

echasnovski commented:
> The git rev is just 7 characters.
> 
> Since this file is not intended to be human edited should we use the the full git hash ?
> 
> This also prevents the (probably) rare situation where 7 characters of a git hash could mean more than 1 commit.

A note: it is not restricted to 7 characters, is uses `--abbrev-commit` which should automatically decide which width is appropriate for the repo. For example, 'nvim-treesitter/nvim-treesitter' already uses 8 characters. It also can depend on Git config.

I though about using full hash here, which is definitely the safer option. Here is the reasoning of why it doesn't:
- It looks like a good idea to use the same revision string in both confirmation buffer and lockfile. The confirmation buffer uses abbreviated revision because it looks less intimidating to users.
- The lockfile is recommended now to be used for freezing to a particular commit. Shorter strings look nicer in config.
- It will add extra moving part to the code and tests.

If none of these concerns are reasonable (cc @justinmk), then changing `vim.pack` to *always* use full hash to indicate revision seems like a better idea.

justinmk commented:
> If none of these concerns are reasonable (cc @justinmk), then changing `vim.pack` to _always_ use full hash to indicate revision seems like a better idea.

I would suggest at least 12 chars, that was discussed on the git mailing list at some point as the sweet spot.
